File: Eyes/ImageTester/src/csharp/lib/TestObjects/Suite.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Applitools.Images;

namespace Applitools.ImageTester.TestObjects
{
    public class Suite : TestUnit
    {
        private IList<Batch> batches;
        private Test test_ = null;

        public Suite(FileStream file) : base(file)
        {
            batches = new List<Batch>();
        }

        public override void Run(Eyes eyes)
        {
            if (!batches.Any() || test_ == null)
            {
                Console.WriteLine("Nothing to test!");
                return;
            }

            foreach (Batch batch in batches)
            {
                try
                {
                    batch.Run(eyes);
                }
                finally
                {
                    batch.Dispose();
                }
            }

            Console.WriteLine("> No batch <");
            test_.Run(eyes);
        }

        public virtual void AddBatch(Batch batch)
        {
            batches.Add(batch);
        }

        public virtual void PortBatchesTo(Suite destination)
        {
            destination.batches = destination.batches.Concat(batches).ToList();
            batches.Clear();
        }

        public virtual void PortTestTo(Batch destination)
        {
            destination.AddTest(test_);
            test_ = null;
        }

        public virtual bool HasOrphanTest()
        {
            return test_ != null;
        }

        public virtual void SetTest(Test test)
        {
            if (test_ != null)
            {
                throw new ArgumentException("test is not null as expected!");
            }

            test_ = test;
        }

        public override void Dispose()
        {
            test_?.Dispose();
        }
    }
}

File: Eyes/ImageTester/src/lib/java/com/applitools/ImageTester/TestObjects/Suite.java
package com.applitools.ImageTester.TestObjects;


import com.applitools.eyes.images.Eyes;

import java.io.File;
import java.io.IOException;
import java.util.LinkedList;
import java.util.Queue;

public class Suite extends TestUnit {
    private Queue<Batch> batches_;
    private Test test_ = null;

    public Suite(File file) {
        super(file);
        batches_ = new LinkedList<Batch>();
    }

    public void run(Eyes eyes) throws IOException {
        if (batches_.isEmpty() && test_ == null) {
            System.out.printf("Nothing to test!\n");
            return;
        }

        for (Batch batch : batches_) {
            try {
                batch.run(eyes);
            }finally {
                batch.dispose();
            }
        }

        if (test_ != null) {
            System.out.printf("> No batch <\n");
            test_.run(eyes);
        }
    }

    public void addBatch(Batch batch) {
        batches_.add(batch);
    }

    public void portBatchesTo(Suite destination) {
        destination.batches_.addAll(batches_);
        batches_.clear();
    }

    public void portTestTo(Batch destination) {
        destination.addTest(test_);
        test_ = null;
    }

    public boolean hasOrphanTest() {
        return test_ != null;
    }

    public void setTest(Test test) {
        if (test_ != null) throw new RuntimeException("test is not null as expected!");
        test_ = test;
    }

    public void dispose() {
        if (test_ != null) test_.dispose();
    }
}
